Wiki - Assign built in policy (#297)
Azure/terraform-azurerm-caf-enterprise-scale: docs/wiki/[Examples]-Assign-a-Built-in-Policy.md

## Overview

> This page describes how to assign built-in Azure Policies to your Enterprise Scale deployment.

Reviewer: ```suggestion
This page describes how to assign built-in Azure Policies to your environment using the `caf-enterprise-scale` module.
```

> In this example you will use two built-in policies and one built-in policy set definition. You will use policies that are not already availble for assignment via the default  Enterprise Scale deployment.

Reviewer: ```suggestion
In this example you will use two built-in policies and one built-in policy set definition. You will use policies that are not already available for assignment via the default Enterprise Scale deployment.

```

> The policies you will use are `Not allowed resource types`  and `Deploy default Microsoft IaaSAntimalware extension for Windows Server`. The policy set definition (Initiative) you will use is  `NIST SP 800-53 Rev. 5`.

Reviewer: ```suggestion
The policies you will use are `Not allowed resource types` and `Deploy default Microsoft IaaSAntimalware extension for Windows Server`. 

The policy set definition (Initiative) you will use is `NIST SP 800-53 Rev. 5`.
```

- Assign the policy definition for `Deploy default Microsoft IaaSAntimalware extension for Windows Server` at the `es_root` Management Group by extending the built-in archetype for `es_root`
- Assign the policy set definition for `NIST SP 800-53 Rev. 5` at the `es_root` Management Group by extending the built-in archetype for `es_root`
- Assign the policy definition for `Not allowed resource types` at the `Landing Zones` Management Group by extending the built-in archetype for `es_landing_zones`.

>IMPORTANT: To allow the declaration of custom or expanded templates, you must create a custom library folder within the root module and include the path to this folder using the `library_path` variable within the module configuration. In our example, the directory is `/lib`.

In order to assign built-in policies, there needs to be an assignment file for each policy or policy set definition that we want to use.

> This module already includes assignment files for some built-in policies so it's important to check whether or not one exists before creating your own. You can do this by navigating to  `\modules\archetypes\lib\policy_assignments` and looking for an assignment file that matches the policy you want to assign.

Reviewer: Provide a link to the lib https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/tree/main/modules/archetypes/lib

Author: Thanks Jack. I'll add this in the next commit.

> An example of a built-in policy that already has an assignment file included within the module is the `Kubernetes clusters should be accessible only over HTTPS` policy. The assignment file for this policy is called `policy_assignment_es_deny_http_ingress_aks.tmpl.json`.

Reviewer: again provide link to this file to help them get tangible with it

Author: Thanks Jack. I'll add this in the next commit.

As the policies you will use in this example do not already have an assignment file within the module, you will need to create the below files so that you can assign them:

- [lib/policy_assignments/policy_assignment_not_allowed_resource_types.json](#libpolicy_assignmentspolicy_assignment_not_allowed_resource_typesjson)
- [lib/policy_assignments/policy_assignment_deploy_default_microsoft_IaaSAntimalware_extension_for_windows_server.json](#libpolicy_assignmentspolicy_assignment_deploy_default_microsoft_IaaSAntimalware_extension_for_windows_serverjson)
- [lib/policy_assignments/policy_assignment_nist_sp_800_53_rev_5.json](#libpolicy_assignmentspolicy_assignment_nist_sp_800_53_rev_5json)

## Create Custom Policy Assignment Files

In order to assign built-in policies or policy sets, you need to create policy assignment files. The first step is to create a `policy_assignments` subdirectory within `/lib`.

>**NOTE**: Creating a `policy_assignments` subdirectory within `\lib` is a recommendation only. If you prefer not to create one or to call it something else, the custom policies will still work.

> You will then need to create a file named `policy_assignment_not_allowed_resource_types.json` within the `policy_assignments` directory. Copy the below code in to the file and save it.

Reviewer: add another note saying the file name is not important, but the key in the JSON file is. As per the https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/wiki/%5BUser-Guide%5D-Archetype-Definitions doc

>**NOTE**: The `name` parameter in the assignment file has an upper limit of 24 characters. Values longer than this will result in a validation error. As an example, `"Not allowed resources` fits but `"Not allowed resource types"` would result in an error.

> To find the appropriate value for `policyDefinitionID` you will need to locate the policy you want to use in the portal and copy the `Definition ID` value.

Reviewer: provide link to policy PG docs about finding definition id if available, like via PWSH or AZ CLI

Author: Added links to PowerShell and CLI docs.

      "access_control": {
      }
    }
  }
}
```

You should now kick-off your Terraform workflow (init, plan, apply) to apply the new configuration. This can be done either locally or through a pipeline. When your workflow has finished, the `NIST SP 800-53 Rev. 5` policy set will be assigned at `es_root`.

## Assign the `Not allowed resource types` Policy at the `Landing Zones` Management Group

In this example, we will assign it at the `Landing Zones` Management Group. To do this, either update your existing `archetype_extension_es_landing_zones.tmpl.json` file or create one and copy the below code in to it and save.

```json
{
  "extend_es_landing_zones": {
    "policy_assignments": ["Not allowed resource types"],
    "policy_definitions": [],
    "policy_set_definitions": [],
    "role_definitions": [],
    "archetype_config": {
      "access_control": {
      }
    }
  }
}
```

You should now kick-off your Terraform workflow once again to apply the updated configuration. This can be done either locally or through a pipeline. When your workflow has finished, the `Not allowed resource types` policy will be assigned at the `Landing Zones` Management Group.

> ```hcl

Reviewer: remove this as not needed

Author: Removed the HCL code for 'terraform apply'

- Assign the policy set definition for `NIST SP 800-53 Rev. 5` at the `es_root` Management Group by extending the built-in archetype for `es_root`
- Assign the policy definition for `Not allowed resource types` at the `Landing Zones` Management Group by extending the built-in archetype for `es_landing_zones`.

>**IMPORTANT**: To allow the declaration of custom or expanded templates, you must create a custom library folder within the root module and include the path to this folder using the `library_path` variable within the module configuration. In our example, the directory is `/lib`.

In order to assign built-in policies, there needs to be an assignment file for each policy or policy set definition that we want to use.
This module already includes assignment files for some built-in policies so it's important to check whether or not one exists before creating your own.

> You can do this by navigating to  [`\modules\archetypes\lib\policy_assignments`](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/tree/main/modules/archetypes/lib) and looking for an assignment file that matches the policy you want to assign.

Reviewer: ```suggestion
You can do this by navigating to  [`\modules\archetypes\lib\policy_assignments`](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/tree/main/modules/archetypes/lib) and looking for an assignment file that matches the policy you want to assign.

```
